### PR TITLE
Compliance with the xSDK community policies [xsdk-cpissues-dev]

### LINKED ISSUE #275: MFEM in xSDK
Note: This is a mirror of https://github.com/xsdk-project/xsdk-issues/issues/11

MFEM is joining the xSDK-0.3 release, which requires that we conform to a set of [community policies](http://dx.doi.org/10.6084/m9.figshare.4495136). Our current policy gaps are:

- [x] **M1**: xSDK-specific options in CMake, xsdk-project/xsdk-issues#15 
- [x] **M3**: user-provided MPI communicators, xsdk-project/xsdk-issues#16 
- [x] **M8**: runtime API for version number, xsdk-project/xsdk-issues#37 
- [x] **M11**: disable output/redirect stdout, xsdk-project/xsdk-issues#18
- [x] **M16**: Spack support, xsdk-project/xsdk-issues#19

Discuss these policies, our compliance, and progress below.

### LINKED ISSUE #333: Incompatibility with SUNDIALS 3.0.0
I tried to compile with the new SUNDIALS release 3.0.0 and recognized an incompatibility.

It should be mentioned somewhere that we are compatible up to version 2.7.0

### PR DESCRIPTION
**Pull request by acfisher**
This branch addresses the changes needed to bring us in to compliance with the XSDK community policy.

Resolves #275.
Resolves #333.

TODO:
- [ ] Document the changes for SUNDIALS 3.0.0 in the `CHANGELOG` (and maybe `INSTALL`) files
- [ ] Fix single/double quote issues in miapps "make test"

### CODE REVIEW

File: general/communication.hpp [hunk 1]
```diff
@@ -19,31 +19,30 @@
...
 #include <mpi.h>
 
+
 namespace mfem
 {
 
 /** @brief A simple convenience class that calls MPI_Init() at construction and
     MPI_Finalize() at destruction. It also provides easy access to
-    MPI_COMM_WORLD's rank and size. */
+    to the global_mpi_comm rank and size. */
```
Comment (tzanio) [inline general/communication.hpp]: `MFEM_COMM_WORLD` instead of `global_mpi_comm`?

File: general/communication.hpp [hunk 2]
```diff
@@ -19,31 +19,30 @@
...
-      MPI_Comm_size(MPI_COMM_WORLD, &world_size);
-   }
+   void GetRankAndSize();
 public:
    MPI_Session() { MPI_Init(NULL, NULL); GetRankAndSize(); }
    MPI_Session(int &argc, char **&argv)
    { MPI_Init(&argc, &argv); GetRankAndSize(); }
    ~MPI_Session() { MPI_Finalize(); }
-   /// Return MPI_COMM_WORLD's rank.
+   /// Return global_mpi_comm's rank.
```
Comment (tzanio) [inline general/communication.hpp]: `MFEM_COMM_WORLD` instead of `global_mpi_comm`?

File: general/communication.hpp [hunk 3]
```diff
@@ -19,31 +19,30 @@
...
    int WorldRank() const { return world_rank; }
-   /// Return MPI_COMM_WORLD's size.
+   /// Return global_mpi_comm's size.
```
Comment (tzanio) [inline general/communication.hpp]: `MFEM_COMM_WORLD` instead of `global_mpi_comm`?

File: general/version.hpp [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+// terms of the GNU Lesser General Public License (as published by the Free
+// Software Foundation) version 2.1 dated February 1999.
+
+#ifndef MFEM_VERSION_H
+#define MFEM_VERSION_H
+
+namespace mfem
+{
+
+int get_version();
```
Comment (tzanio) [inline general/version.hpp]: Switch to `GetVersion()` instead of `get_version()`? Seems more mfem-y

File: general/globalcomm.hpp [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+// terms of the GNU Lesser General Public License (as published by the Free
+// Software Foundation) version 2.1 dated February 1999.
+
+#ifndef MFEM_GLOBAL_COMM
+#define MFEM_GLOBAL_COMM
+
+#include "../config/config.hpp"
+#include <mpi.h>
+
+#ifdef MFEM_USE_MPI
```
Comment (tzanio) [inline general/globalcomm.hpp]: Move before `#include <mpi.h>`

File: general/globalcomm.hpp [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+namespace mfem
+{
+
+extern MPI_Comm MFEM_COMM_WORLD;
+
+}
+
+#endif
+
+#endif
```
Comment (tzanio) [inline general/globalcomm.hpp]: Add newline

File: fem/gridfunc.cpp [hunk 7]
```diff
@@ -2452,10 +2452,10 @@ void GridFunction::SaveSTL(std::ostream &out, int TimesToRefine)
...
    }
 
-   cout << "[xmin,xmax] = [" << bbox[0][0] << ',' << bbox[0][1] << "]\n"
-        << "[ymin,ymax] = [" << bbox[1][0] << ',' << bbox[1][1] << "]\n"
-        << "[zmin,zmax] = [" << bbox[2][0] << ',' << bbox[2][1] << ']'
-        << endl;
+   out << "[xmin,xmax] = [" << bbox[0][0] << ',' << bbox[0][1] << "]\n"
+       << "[ymin,ymax] = [" << bbox[1][0] << ',' << bbox[1][1] << "]\n"
+       << "[zmin,zmax] = [" << bbox[2][0] << ',' << bbox[2][1] << ']'
+       << endl;
```
Comment (v-dobrev) [inline fem/gridfunc.cpp]: Since 'out' is the name of a parameter, this should be `mfem::out`.

Comment (acfisher) [reply fem/gridfunc.cpp]: Yeah, I saw that one.  Do we really want the output to potentially go to 2 different places?

Comment (v-dobrev) [reply fem/gridfunc.cpp]: Yes, this is just informational message and it is not part of the STL file that we are writing to `out`. I'll fix this, with some other changes I'm making.

File: general/globalostream.hpp [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+   inline bool IsEnabled() {return enabled;}
+};
+
+
+inline std::ostream& WrappedOStream::GetStream()
+{
+#ifdef MFEM_USE_MPI
+   int world_rank;
+   MPI_Comm_rank(MFEM_COMM_WORLD, &world_rank);
+   if (enabled && world_rank == 0)
```
Comment (v-dobrev) [inline general/globalostream.hpp]: I think the default should be for all ranks to keep the output enabled.

Any objections to change that?

Comment (tzanio) [reply general/globalostream.hpp]: I personally prefer printing only on rank 0, but I'm flexible on that. The current choice seems to be also more in the spirit of M11, which says:

> Also, packages may print to stdout by default but only on one process (i.e., root rank “0”).

File: general/communication.hpp [hunk 9]
```diff
@@ -19,31 +19,30 @@
...
-   {
-      MPI_Comm_rank(MPI_COMM_WORLD, &world_rank);
-      MPI_Comm_size(MPI_COMM_WORLD, &world_size);
-   }
+   void GetRankAndSize(MPI_Comm);
 public:
-   MPI_Session() { MPI_Init(NULL, NULL); GetRankAndSize(); }
-   MPI_Session(int &argc, char **&argv)
-   { MPI_Init(&argc, &argv); GetRankAndSize(); }
+   MPI_Session(MPI_Comm world_comm = MPI_COMM_WORLD)
```
Comment (v-dobrev) [inline general/communication.hpp]: Looking at this again, I don't think any communicator other than `MPI_COMM_WORLD` makes sense in this class - it is meant to initialize MPI, so you cannot have any other communicators.

We should probably just remove this optional parameter.

File: general/error.cpp [hunk 10]
```diff
@@ -10,7 +10,9 @@
 // Software Foundation) version 2.1 dated February 1999.
 
 #include "error.hpp"
+#include "globals.hpp"
 #include "array.hpp"
+#include "communication.hpp"
```
Comment (v-dobrev) [inline general/error.cpp]: No need to include this here, I think.

File: general/version.hpp (hunk 4 above)
Comment (v-dobrev) [reply general/version.hpp]: We do use "snake_case" for some functions, e.g. mfem_error(), and some classes, e.g. socketstream. I'm OK with both.

File: mesh/mesquite.cpp [hunk 11]
```diff
@@ -715,7 +715,7 @@ void MesquiteMesh::elements_get_topologies(const ElementHandle
 
 MesquiteMesh::~MesquiteMesh()
 {
-   MsqPrintError err(cerr);
+   MsqPrintError err(mfem::err.GetStream());
```
Comment (v-dobrev) [inline mesh/mesquite.cpp]: Oops, need to update this: `mfem::err.GetStream() -> mfem::err`.

File: mesh/mesquite.cpp [hunk 12]
```diff
@@ -847,7 +847,7 @@ static void BoundaryPreservingOptimization(mfem::MesquiteMesh &mesh)
 {
 
    MsqDebug::enable(1);
-   MsqPrintError err(cerr);
+   MsqPrintError err(mfem::err.GetStream());
```
Comment (v-dobrev) [inline mesh/mesquite.cpp]: Replace `mfem::err.GetStream() -> mfem::err`.

File: mesh/mesquite.cpp [hunk 13]
```diff
@@ -1145,7 +1145,7 @@ void mfem::Mesh::MesquiteSmooth(const int mesquite_option)
 {
    mfem::MesquiteMesh msq_mesh(this);
    MsqDebug::enable(1);
-   MsqPrintError err(cerr);
+   MsqPrintError err(mfem::err.GetStream());
```
Comment (v-dobrev) [inline mesh/mesquite.cpp]: Replace `mfem::err.GetStream() -> mfem::err`.

File: general/communication.hpp (hunk 9 above)
Comment (tzanio) [reply general/communication.hpp]: Agreed

File: general/communication.hpp (hunk 3 above)
Comment (acfisher) [reply general/communication.hpp]: Done in Veselin's update.

File: general/communication.hpp (hunk 2 above)
Comment (acfisher) [reply general/communication.hpp]: Done in Veselin's update.

File: general/communication.hpp (hunk 1 above)
Comment (acfisher) [reply general/communication.hpp]: Done in Veselin's update.

File: fem/gridfunc.cpp (hunk 7 above)
Comment (acfisher) [reply fem/gridfunc.cpp]: Done in Veselin's update.

File: general/globalcomm.hpp (hunk 5 above)
Comment (acfisher) [reply general/globalcomm.hpp]: Done in Veselin's update.

File: general/globalostream.hpp (hunk 8 above)
Comment (acfisher) [reply general/globalostream.hpp]: Updated to print to all ranks in Veselin's update.

File: general/version.hpp (hunk 4 above)
Comment (acfisher) [reply general/version.hpp]: Changing to camel case in my upcoming update.

File: mesh/mesquite.cpp (hunk 11 above)
Comment (acfisher) [reply mesh/mesquite.cpp]: Done in Veselin's update.

File: mesh/mesquite.cpp (hunk 12 above)
Comment (acfisher) [reply mesh/mesquite.cpp]: Done in Veselin's update.

File: mesh/mesquite.cpp (hunk 13 above)
Comment (acfisher) [reply mesh/mesquite.cpp]: Done in Veselin's update.

**Comment by acfisher** [comment]
Thank you Veselin for addressing most of the changes.  I just put in the change of get_version to camelcase, and ran a few tests to make sure everything still works.  I believe all of the comments have been address at this point.

**Comment by acfisher** [comment]
Veselin, are you happy with the branch in it's current state?  If so I'll put the xsdk rc2 tag on it so the xsdk folks will use this state for their initial testing.

**Comment by v-dobrev** [comment]
Yes, I'm happy with the current state. I wanted to wait for feedback and any corrections/additions before final approval.

**Comment by tzanio** [comment]
Yes, this is definitely ready for the `rc2` tag.

I also think we can merge into `next` and re-merge later if necessary,

File: miniapps/electromagnetics/makefile [hunk 14]
```diff
@@ -59,10 +59,10 @@ include $(MFEM_TEST_MK)
 RUN_MPI = $(MFEM_MPIEXEC) $(MFEM_MPIEXEC_NP) $(MFEM_MPI_NP)
 volta-test-par: volta
 	@$(call mfem-test,$<, $(RUN_MPI), Electromagnetic miniapp,\
-	-maxit 2 -dbcs 1 -dbcg -ds '0.0 0.0 0.0 0.2 8.0')
```
Comment (tzanio) [inline miniapps/electromagnetics/makefile]: What's the reason for this change? If I remember correctly single quotes were necessary for srun on LC...

Comment (acfisher) [reply miniapps/electromagnetics/makefile]: It was failing on rzalastor with single quotes.  I sure hope we're not in a situation where we need different quotes on different platforms.

Comment (tzanio) [reply miniapps/electromagnetics/makefile]: @aaroncblack, do you remember the issue/reason for the quotes?

**Comment by tzanio** [comment]
Added two TODO items in the PR description.
